pref-adjust.lua

```diff
@@ -1,80 +1,165 @@
+    detests trolls, buzzards, vultures and crundles.
 
-]====]
+Additionally, ``pref-adjust goth`` will insert a less than ideal set, which
+is quite challenging, for a single dwarf::
+
+    ... likes dwarf skin, corpses, body parts, remains, coffins, the color
+    black, crosses, glumprongs for their living shadows and snow demons for
+    their horrifying features.  When possible, she prefers to consume sewer
+    brew, gutter cruor and bloated tubers.  She absolutely detests elves,
+    humans and dwarves.
+
+To see what values can be used with each type of preference, use
+``pref-adjust list``.  Optionally, a single dwarf or all dwarves can have
+their preferences cleared manually with the use of ``pref-adjust clear_one``
+and ``pref-adjust clear_all``, respectively. Existing preferences are
+automatically cleared, normally.
 
+]====]
 -- ---------------------------------------------------------------------------
-function brainwash_unit(unit)
+function insert_preference(unit, preftype, val1)
+
+    if preftype == df.unit_preference.T_type.LikeMaterial then
+        utils.insert_or_update(unit.status.current_soul.preferences, {
+            new = true,
+            type = preftype,
+            item_type = -1,
+            poetic_form_id = -1,
+            musical_form_id = -1,
+            dance_form_id = -1,
+            mattype = dfhack.matinfo.find(val1).type,
+            mat_state = df.matter_state.Solid,
+            matindex = dfhack.matinfo.find(val1).index,
+            active = true,
+            prefstring_seed = pss_counter,
+        }, 'prefstring_seed')
+        -- mattype for some is non zero, those non-inorganic like creature:gazelle:hoof is 42,344
+    elseif preftype == df.unit_preference.T_type.LikeFood then
+        consumable_type = val1[1]
+        consumable_name = val1[2]
+        utils.insert_or_update(unit.status.current_soul.preferences, {
+            new = true,
+            type = preftype,
+            item_type = consumable_type,
+            poetic_form_id = consumable_type,
+            musical_form_id = df.consumable_type,
+            dance_form_id = consumable_type,
+            item_subtype = dfhack.matinfo.find(consumable_name).subtype,
+            mattype = dfhack.matinfo.find(consumable_name).type,
+            mat_state = df.matter_state.Solid,
+            matindex = dfhack.matinfo.find(consumable_name).index,
+            active = true,
+            prefstring_seed = pss_counter,
+        }, 'prefstring_seed')
+    elseif df.unit_preference.T_type[preftype] ~= nil then
+        utils.insert_or_update(unit.status.current_soul.preferences, {
+            new = true,
+            type = preftype,
+            item_type = val1,
+            creature_id = val1,
+            color_id = val1,
+            shape_id = val1,
+            plant_id = val1,
+            poetic_form_id = val1,
+            musical_form_id = val1,
+            dance_form_id = val1,
```

Comment:
These 8 fields are a union: https://github.com/DFHack/df-structures/blob/0.47.04-r1/df.units.xml#L2184-L2193
Essentially, they all refer to the same location in memory, just under different names, so you only need to set one of them. (I assume you did it this way because you noticed that they always have the same value - that's true, but you only need to set one of them to make that happen.)

Reply:
It's actually not *quite* as simple as I made it out to be - only the last 3 fields are int32s, so setting any of the first 5 fields will only set the lower two bytes of the last 3. `assign()` also only appears to assign the first field passed to it that's in the union (i.e. only the value you give for `item_type` matters here), so I think the safest option here (albeit confusing) is to assign to `poetic_form_id` only. I'll leave a comment to that effect.